TrafficCapture/coreUtilities/src/main/java/org/opensearch/migrations/tracing/RootOtelContext.java

```diff
@@ -0,0 +1,139 @@
+                        // "A Prometheus Exporter MUST only support Cumulative Temporality."
+                        //.setAggregationTemporalitySelector(AggregationTemporalitySelector.deltaPreferred())
+                        .build())
+                .setInterval(Duration.ofMillis(1000))
+                .build();
+        final var logProcessor = BatchLogRecordProcessor.builder(OtlpGrpcLogRecordExporter.builder()
+                        .setEndpoint(collectorEndpoint)
+                        .build())
+                .build();
+
+        var openTelemetrySdk = OpenTelemetrySdk.builder()
+                .setTracerProvider(SdkTracerProvider.builder().setResource(serviceResource)
+                        .addSpanProcessor(spanProcessor).build())
+                .setMeterProvider(SdkMeterProvider.builder().setResource(serviceResource)
+                        .registerMetricReader(metricReader).build())
+                .setLoggerProvider(SdkLoggerProvider.builder().setResource(serviceResource)
+                        .addLogRecordProcessor(logProcessor).build())
+                .build();
+
+        // Add hook to close SDK, which flushes logs
+        Runtime.getRuntime().addShutdownHook(new Thread(openTelemetrySdk::close));
+        return openTelemetrySdk;
+    }
+
+    public static OpenTelemetry initializeOpenTelemetry(String collectorEndpoint, String serviceName) {
+        return Optional.ofNullable(collectorEndpoint)
+                .map(endpoint -> initializeOpenTelemetryForCollector(endpoint, serviceName))
+                .orElse(OpenTelemetrySdk.builder().build());
+    }
+
+
+    public RootOtelContext(String scopeName) {
+        this(scopeName, null);
+    }
+
+    public RootOtelContext(String scopeName, String collectorEndpoint, String serviceName) {
+        this(scopeName, initializeOpenTelemetry(collectorEndpoint, serviceName));
+    }
+
+    public RootOtelContext(String scopeName, OpenTelemetry sdk) {
+        openTelemetryImpl = sdk != null ? sdk : initializeOpenTelemetry(null, null);
+        this.scopeName = scopeName;
+    }
+
+    @Override
+    public RootOtelContext getEnclosingScope() {
+        return null;
+    }
+
+    OpenTelemetry getOpenTelemetry() {
+        return openTelemetryImpl;
+    }
+
+    @Override
+    public MeterProvider getMeterProvider() {
+        return getOpenTelemetry().getMeterProvider();
+    }
+
+    @Override
+    public AttributesBuilder fillAttributes(AttributesBuilder builder) {
+        return builder; // nothing more to do
+    }
+
+    private static SpanBuilder addLinkedToBuilder(Stream<Span> linkedSpanContexts, SpanBuilder spanBuilder) {
+        return Optional.ofNullable(linkedSpanContexts)
+                .map(ss->ss.collect(Utils.foldLeft(spanBuilder, (b,s)->b.addLink(s.getSpanContext()))))
+                .orElse(spanBuilder);
+    }
+
+    private static Span buildSpanWithParent(SpanBuilder builder, Attributes attrs, Span parentSpan,
```

Comment:
This was pretty hard to understand and read the order of calls on the builder. 

I'd prefer this, note as well a rename of "linkedSpanContexts" to "linkedSpans" to align with the type

```java
    private static SpanBuilder addLinkedSpansToBuilder(Stream<Span> linkedSpans, SpanBuilder spanBuilder) {
        if (linkedSpans != null) {
            linkedSpans.map(Span::getSpanContext).forEach(spanBuilder::addLink);
        }
        return spanBuilder;
    }

    private static SpanBuilder addParentToBuilder(Span parentSpan, SpanBuilder spanBuilder) {
        return parentSpan != null
            ? spanBuilder.setParent(Context.current().with(parentSpan))
            : spanBuilder.setNoParent();
    }

    private static Span buildSpanWithParent(SpanBuilder builder, Attributes attrs, Span parentSpan,
                                            Stream<Span> linkedSpans) {
        addParentToBuilder(parentSpan, builder);
        addLinkedSpansToBuilder(linkedSpans, builder);
        return builder.startSpan().setAllAttributes(attrs);
    }
```

Reply:
The code above will only work if the builder is a mutable builder.  The way that the code is written now, the builder could be mutable or immutable since the newly returned builder is returned as input into the next call.  I could use the javadocs to see that 'this' is returned, but that requires the programmer to remember that convention and makes it harder to do audits over the code.  Looking to rebind the "current builder" as  the newly returned builder value is what makes it difficult to call two functions in sequence (rather than compositionally).

The middle ternary could replace the monadic ofNullable, but I prefer the ofNullable because it creates a newer & tighter scope for the nonNullable value that we're operating on and fences values my from leaving their scope during refactorings.